firestore-cache/src/main.js

```diff
@@ -0,0 +1,173 @@
+// [[start common_connector_code]]
+var cc = DataStudioApp.createCommunityConnector();
+
+function getAuthType() {
+  return cc
+    .newAuthTypeResponse()
+    .setAuthType(cc.AuthType.NONE)
+    .build();
+}
+
+function getConfig(request) {
+  var config = cc.getConfig();
+
+  config
+    .newTextInput()
+    .setId('zipcode')
+    .setName('Enter Zip Code')
+    .setPlaceholder('eg. 95054');
+
+  return config.build();
+}
+
+function getFields() {
+  var fields = cc.getFields();
+  var types = cc.FieldType;
+  var aggregations = cc.AggregationType;
+
+  fields
+    .newDimension()
+    .setId('zipcode')
+    .setName('Zip code')
+    .setType(types.TEXT);
+
+  fields
+    .newDimension()
+    .setId('date')
+    .setName('Date')
+    .setType(types.YEAR_MONTH_DAY);
+
+  fields
+    .newMetric()
+    .setId('temperature')
+    .setName('Temperature (F)')
+    .setType(types.NUMBER)
+    .setIsReaggregatable(false);
+
+  return fields;
+}
+
+function getSchema(request) {
+  return {
+    schema: getFields().build()
+  };
+}
+// [[end common_connector_code]]
+
+// [[start caching_implementation]]
+function getData(request) {
+  var requestedFields = getFields().forIds(
+    request.fields.map(function(field) {
+      return field.name;
+    })
+  );
+
+  var cacheUpdateNeeded = true;
+  var url = buildFirebaseUrl(request.configParams.zipcode);
+  var cache = firebaseCache('get', url);
+
+  if (cache) {
+    var currentYmd = getCurrentYmd();
+    cacheUpdateNeeded = currentYmd > cache.ymd;
+  }
+
+  if (cacheUpdateNeeded) {
+    var fetchedData = fetchAndParseData(request);
+    cache = {};
+    cache.data = fetchedData;
+    cache.ymd = currentYmd;
+    firebaseCache('delete', url);
+    firebaseCache('post', url, cache);
+  }
+
+  var data = getFormattedData(cache.data, requestedFields);
+
+  return {
+    schema: requestedFields.build(),
+    rows: data
+  };
+}
+
+function getCurrentYmd() {
+  var currentDate = new Date();
+  var year = currentDate.getFullYear();
+  var month = ('0' + (currentDate.getMonth() + 1)).slice(-2);
+  var date = ('0' + currentDate.getDate()).slice(-2);
+  var currentYmd = year + month + date;
+  return currentYmd;
+}
+
+// [[end caching_implementation]]
+
+// [[start common_getdata_implementation]]
+function fetchAndParseData(request) {
+  // You will connect to your own API endpoint and parse the fetched data.
```

Comment:
nit: Add a `TODO` to the beginning of this comment to draw more attention to it.